controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1153,6 +1179,227 @@ func saveHostProvisioningSettings(host *metal3v1alpha1.BareMetalHost) (dirty boo
 	return
 }
 
+// Get the stored firmware settings and validate the Spec settings against the schema
+func (r *BareMetalHostReconciler) getHostFirmwareSettings(prov provisioner.Provisioner, info *reconcileInfo) (hfs *metal3v1alpha1.HostFirmwareSettings, err error) {
+
+	hfs = &metal3v1alpha1.HostFirmwareSettings{}
+	if err = r.Get(context.TODO(), info.request.NamespacedName, hfs); err != nil {
+
+		if !k8serrors.IsNotFound(err) {
+			// Error reading the object
+			return nil, errors.Wrap(err, "could not load host firmware settings")
+		}
+
+		// Could not get settings, log it but don't return error as settings may not have been available at provisioner
+		info.log.Info("could not get hostFirmwareSettings", "namespacename", info.request.NamespacedName)
+		return nil, nil
+	}
+
+	if hfs.Status.Settings == nil {
+		// If settings cleared after pivot, restore the settings
+		currentSettings, _, err := prov.GetFirmwareSettings(false)
+		if err != nil {
+			info.log.Info("could not get firmware settings from provisioner")
+			return nil, err
+		}
+
+		if err = r.updateHostFirmwareSettingsFromCurrent(currentSettings, hfs, nil); err != nil {
+			return nil, err
+		}
+	}
+
+	schema := &metal3v1alpha1.FirmwareSchema{}
+	if hfs.Status.FirmwareSchema != nil {
+		if err = r.Get(context.TODO(), client.ObjectKey{
+			Namespace: hfs.Status.FirmwareSchema.Namespace,
+			Name:      hfs.Status.FirmwareSchema.Name}, schema); err != nil {
+
+			info.log.Error(err, "failed to get schema for hostFirmwareSettings")
+			return nil, err
+		}
+	} else {
+		schema = nil
+	}
+
+	// Validate the Spec setting using schema, if available
+	if err = ValidateHostFirmwareSettings(hfs, schema); err != nil {
+		return nil, err
+	}
+
+	return hfs, nil
+}
+
+// Get the firmware settings from the provisioner and create hostFirmwareSettings and FirmwareSchema resources
+func (r *BareMetalHostReconciler) storeHostFirmwareSettings(prov provisioner.Provisioner, info *reconcileInfo) (err error) {
+
+	info.log.Info("storeHostFirmwareSettings")
+
+	// First check if there is a hostFirmwareSettings resource for this host
+	hfs := &metal3v1alpha1.HostFirmwareSettings{}
+	if err = r.Get(context.TODO(), info.request.NamespacedName, hfs); err == nil {
+		info.log.Info("found existing hostFirmwareSettings resource")
+	} else if !k8serrors.IsNotFound(err) {
+		// Error reading the object
+		return errors.Wrap(err, "could not load host firmware settings")
+	} else {
+		// Create a new resource
+		if err = r.newHostFirmwareSettings(info, hfs); err != nil {
```

Comment:
Wouldn't be safer to postpone the resource creation after a successful `prov.GetFirmwareSettings(true)`?